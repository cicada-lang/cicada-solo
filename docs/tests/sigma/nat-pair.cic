let one = add1(zero)
let two = add1(one)
let three = add1(two)
let four = add1(three)

same_as_chart! Type [
  [_: Nat | Nat],
  Pair(Nat, Nat),
]

same_as_chart! [_: Nat | Nat] [
  cons(one, two),
  [one | two],
]

let one_two: [_: Nat | Nat] = cons(one, two)

same_as_chart! Nat [
  cdr(one_two),
  two,
]

same_as_chart! Type [
  [_: Nat, _: Nat | Nat],
  Pair(Nat, Pair(Nat, Nat)),
]

same_as_chart! [_: Nat, _: Nat | Nat] [
  cons(one, cons(two, three)),
  [one, two, | three],
]

let one_two_three: [_: Nat, _: Nat | Nat] = cons(one, cons(two, three))

same_as_chart! Nat [
  cdr(cdr(one_two_three)),
  three,
]

same_as_chart! Type [
  [_: Nat, _: Nat, _: Nat | Nat],
  Pair(Nat, Pair(Nat, Pair(Nat, Nat))),
]

same_as_chart! [_: Nat, _: Nat, _: Nat | Nat] [
  cons(one, cons(two, cons(three, four))),
  [one, two, three | four],
]

let one_two_three_four: [_: Nat, _: Nat, _: Nat | Nat] = cons(one, cons(two, cons(three, four)))

same_as_chart! Nat [
  cdr(cdr(cdr(one_two_three_four))),
  four,
]
