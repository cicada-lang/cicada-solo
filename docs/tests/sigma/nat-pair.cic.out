refl: Equal(Type, [_: Nat | Nat], [_: Nat | Nat])
refl: Equal([_: Nat | Nat], [1 | 2], [1 | 2])
refl: Equal(Nat, 2, 2)
refl: Equal(Type, [_: Nat, _1: Nat | Nat], [_: Nat, _1: Nat | Nat])
refl: Equal([_: Nat, _1: Nat | Nat], [1 | [2 | 3]], [1 | [2 | 3]])
refl: Equal(Nat, 3, 3)
refl: Equal(Type, [_: Nat, _1: Nat, _2: Nat | Nat], [_: Nat, _1: Nat, _2: Nat | Nat])
refl: Equal([_: Nat, _1: Nat, _2: Nat | Nat], [1 | [2 | [3 | 4]]], [1 | [2 | [3 | 4]]])
refl: Equal(Nat, 4, 4)